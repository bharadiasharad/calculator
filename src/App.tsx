import React, { useEffect, useState } from 'react';
import './styles/App.css';
import { Card, FloatButton, Typography } from 'antd';
import { PlusOutlined } from '@ant-design/icons';
import RowList from './components/RowList';
import { Operation } from './common/enum/Operation';
import { RowDataModal } from './modals/RowDataModal';
import { DropdownOption } from './modals/Interface/DropdownOption';

const App: React.FC = () => {

  // State to manage the list of data and the calculated result.
  const [data, setData] = useState<Array<RowDataModal>>([]);
  // State to hold the calculated result.
  const [result, setResult] = useState<number>(0);

  /**
   * Effect hook to recalculate the result whenever the data changes.
   * @param {Array<RowDataModal>} data - The data to monitor for changes.
   */
  useEffect(() => {
    calculateAndSetResult(data);
  }, [data]);

  /**
   * Function to add a new row to the data list.
   */
  const addRow = (): void => {
    setData([...data, new RowDataModal({})]);
  }

  /**
   * Function to remove a row from the data list.
   * @param {number} index - The index of the row to remove.
   */
  const handleRemoveRow = (index: number): void => {
    setData((data) => {
      return [...data].filter(( _ , i) => i !== index )
    })
  };

  /**
   * Function to toggle the visibility of a row.
   * @param {number} index - The index of the row to change visibility.
   */
  const setRowVisibility = (index: number): void => {
    const updatedRows: Array<RowDataModal> = [...data].map((e: RowDataModal, i: number): RowDataModal => {
      return i === index ? { ...e, isVisible: !e.isVisible } : e;
    });
    setData(updatedRows);
  }

  /**
   * Function to handle changes in the value of a row.
   * @param {number | null} value - The new value for the row.
   * @param {number} index - The index of the row to update.
   */
  const handleChange = (value: number | null, index: number) : void => {
    setData((prevData) => {
      const updatedRows = [...prevData];
      updatedRows[index].value = value;
      return updatedRows;
    });
  }

  /**
   * Function to handle changes in the operation type of a row.
   * @param {string} value - The new operation value for the row.
   * @param {number} index - The index of the row to update.
   */
  const handleOperationChange = (value: string, index: number) : void => {
    setData((prevData) => {
      const updatedRows = [...prevData];
      updatedRows[index].operation = value as Operation;
      return updatedRows;
    });
  }

  /**
   * Function to calculate and set the overall result based on the data list.
   * @param {Array<RowDataModal>} updatedRows - The updated list of data rows.
   */
  const calculateAndSetResult = (updatedRows: Array<RowDataModal>) => {
    const result = updatedRows.reduce((total: number, row: RowDataModal) : number => {
      if (!row.value || !row.isVisible) return total;

      switch (row.operation) {
        case Operation.Plus:
          return total + row.value;
        case Operation.Minus:
          return total - row.value;
        default:
          return total
      }
    }, 0);
    setResult(result)
  }

  /**
   * Generate options for the dropdown based on the 'Operation' enum.
   * @type {Array<DropdownOption>}
   */
  const options: Array<DropdownOption> = Object.values(Operation).map((e : string) : DropdownOption => ({ "value": e, "label": e }));

  /**
   * Render the main application component.
   * @returns {JSX.Element} - The JSX element representing the entire application.
   */
  return (
    <div data-testid="test-1" className="card-container">
      <Typography.Title level={2}>Calculator App</Typography.Title>
      <Card data-testid="calculator-card" className="card">
        <FloatButton data-testid="add-row-btn" className="float-button" onClick={() => addRow()} icon={<PlusOutlined />}/>
        <Card>
          <span data-testid="result-span" className="result-text">{result}</span>
        </Card>
        <RowList data={data} handleChange={handleChange} handleOperationChange={handleOperationChange} setRowVisibility={setRowVisibility} handleRemoveRow={handleRemoveRow} options={options} />
      </Card>
    </div>
  );
};

export default App;