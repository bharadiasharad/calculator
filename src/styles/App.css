.card-container {
  padding-left: 20px;
}

.card {
  min-width: 260px;
  width: 360px;
  max-width: 90vw;
  max-height: 85vh;
}

.float-button {
  position: absolute;
  right: 50px;
}

.result-text {
  position: absolute;
  right: 10px;
  bottom: 10px;
}

.list-container {
  min-height: 15vh;
  height: auto;
  max-height: 68vh;
  margin-top: 20px;
  overflow: auto;
  padding-bottom: 60px;
}

.space-compact {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.input-number {
  width: 100%;
}

.icon-button {
  margin-top: 4px;
  padding-left: 12px;
}

.delete-button {
  color: red;
  padding-left: 32px;
  padding-right: 10px;
}

.eye-icon {
  color: blue;
}