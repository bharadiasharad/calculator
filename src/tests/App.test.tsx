import { render, screen, cleanup, fireEvent, waitFor } from "@testing-library/react";
import renderer from 'react-test-renderer';
import App from '../App';

afterEach(() => {
  cleanup();
});

// TODO: write test case to verify different scenarios, disable and deleting the value

test('should render calculator component', () => {
  render(<App />);
  const calculatorCard = screen.getByTestId('calculator-card');
  const addRowButton = screen.getByTestId('add-row-btn');
  const resultSpan = screen.getByTestId('result-span');

  expect(calculatorCard).toBeInTheDocument();
  expect(addRowButton).toBeInTheDocument();
  expect(resultSpan).toBeInTheDocument();
});

test('add 2 rows, assign values, and check the result', async () => {
  render(<App />);
  const addRowButton = screen.getByTestId('add-row-btn');

  fireEvent.click(addRowButton);
  let listItem0;
  await waitFor(() => {
    listItem0 = screen.getByTestId('input-0');
    expect(listItem0).toBeInTheDocument();
  });

  listItem0 && fireEvent.change(listItem0, { target: { value: '2' } });

  fireEvent.click(addRowButton);
  let listItem1;
  await waitFor(() => {
    listItem1 = screen.getByTestId('input-1');
    expect(listItem1).toBeInTheDocument();
  });

  listItem1 && fireEvent.change(listItem1, { target: { value: '3' }});

  const resultComponent = screen.getByTestId('result-span');
  expect(resultComponent).toHaveTextContent('5');
});

test('matches snapshot', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

