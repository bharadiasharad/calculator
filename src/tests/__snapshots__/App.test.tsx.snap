// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="card-container"
  data-testid="test-1"
>
  <h2
    className="ant-typography css-dev-only-do-not-override-amq5gd"
    style={
      Object {
        "WebkitLineClamp": undefined,
      }
    }
  >
    Calculator App
  </h2>
  <div
    className="ant-card ant-card-bordered card css-dev-only-do-not-override-amq5gd"
    data-testid="calculator-card"
    style={Object {}}
  >
    <div
      className="ant-card-body"
      style={Object {}}
    >
      <button
        className="css-dev-only-do-not-override-amq5gd ant-float-btn float-button ant-float-btn-default ant-float-btn-circle"
        data-testid="add-row-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="ant-float-btn-body"
        >
          <div
            className="ant-float-btn-content"
          >
            <div
              className="ant-float-btn-icon"
            >
              <span
                aria-label="plus"
                className="anticon anticon-plus"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                  />
                  <path
                    d="M192 474h672q8 0 8 8v60q0 8-8 8H160q-8 0-8-8v-60q0-8 8-8z"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
      </button>
      <div
        className="ant-card ant-card-bordered css-dev-only-do-not-override-amq5gd"
        style={Object {}}
      >
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <span
            className="result-text"
            data-testid="result-span"
          >
            0
          </span>
        </div>
      </div>
      <div
        className="ant-list ant-list-split list-container css-dev-only-do-not-override-amq5gd"
        style={Object {}}
      >
        <div
          className="ant-spin-nested-loading css-dev-only-do-not-override-amq5gd"
        >
          <div
            className="ant-spin-container"
          >
            <div
              className="ant-list-empty-text"
            >
              <div
                className="css-dev-only-do-not-override-amq5gd ant-empty ant-empty-normal"
                style={Object {}}
              >
                <div
                  className="ant-empty-image"
                >
                  <svg
                    height="41"
                    viewBox="0 0 64 41"
                    width="64"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      fill="none"
                      fillRule="evenodd"
                      transform="translate(0 1)"
                    >
                      <ellipse
                        cx="32"
                        cy="33"
                        fill="#f5f5f5"
                        rx="32"
                        ry="7"
                      />
                      <g
                        fillRule="nonzero"
                        stroke="#d9d9d9"
                      >
                        <path
                          d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                        />
                        <path
                          d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                          fill="#fafafa"
                        />
                      </g>
                    </g>
                  </svg>
                </div>
                <div
                  className="ant-empty-description"
                >
                  No data
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
